### PR TITLE
feat(network): peer manager hadles dial

### PR DESCRIPTION
**Pull request by nagmo-starkware**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

## Other information

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/papyrus/1908)

### CODE REVIEW

**Comment by nagmo-starkware** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/starkware-libs/papyrus/1908?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1908** <a href="https://app.graphite.dev/github/pr/starkware-libs/papyrus/1908?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1905** <a href="https://app.graphite.dev/github/pr/starkware-libs/papyrus/1905?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @nagmo-starkware and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by ShahakShama — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/papyrus/1908)* status: all files reviewed, 12 unresolved discussions (waiting on @nagmo-starkware)

___
*[`crates/papyrus_network/src/peer_manager/behaviour_impl.rs` line 18 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaNKXg6u4Fo0R5gZiC:-NvaNKXg6u4Fo0R5gZiD:bkt45w1) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/behaviour_impl.rs#L18)):*
> ```XML
> pub enum Event {
>     NotifyStreamedBytes(streamed_bytes::behaviour::FromOtherBehaviour),
>     Dial(PeerId, Multiaddr),
> ```

This shouldn't be here. Here are events that reach the other behaviours or the NM
___
*[`crates/papyrus_network/src/peer_manager/behaviour_impl.rs` line 93 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaK_oIBGR7-eNAlP49:-NvaK_oIBGR7-eNAlP4A:b-q9ljva) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/behaviour_impl.rs#L93)):*
> ```XML
>                     error!("Dial failure of an unknow peer. peer id: {}", peer_id)
>                 }
>                 // re-assign a peer to the query otherwise no QueryAssigned event
> ```

Rephrase and capitalise
// Re-assign a peer to the query so that a QueryAssgined Event will be emitted.
___
*[`crates/papyrus_network/src/peer_manager/behaviour_impl.rs` line 96 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaM1dP5VqaHEYvAZ5W:-NvaM1dP5VqaHEYvAZ5X:bk1de13) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/behaviour_impl.rs#L96)):*
> ```XML
>                 // will be emitted.
>                 // TODO: test this case
>                 if let Some((query_id, _)) =
> ```

There might be multiple queries assigned to this peer. You need to re-assign all of them
___
*[`crates/papyrus_network/src/peer_manager/behaviour_impl.rs` line 120 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaMz_8CYsTQ1SAyZze:-NvaMz_8CYsTQ1SAyZzf:b-pf11kz) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/behaviour_impl.rs#L120)):*
> ```XML
>             libp2p::swarm::FromSwarm::ConnectionClosed(ConnectionClosed { peer_id, .. }) => {
>                 if let Some(peer) = self.peers.get_mut(&peer_id) {
>                     peer.set_connection_id(None);
> ```

Check if the connection id of the peer is the closed connection
Also, add a TODO to support multiple connections for a single peer
___
*[`crates/papyrus_network/src/peer_manager/mod.rs` line 106 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaNwbb1k29kxO8z6jh:-NvaNwbb1k29kxO8z6ji:bhqepl2) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/mod.rs#L106)):*
> ```XML
>             );
>             if peer.connection_id().is_none() {
>                 // in case we have a race condition where the connection is closed after we added to
> ```

Capitalisation and punctuation (,.)
___
*[`crates/papyrus_network/src/peer_manager/mod.rs` line 106 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaO69T1NfhxcHUSKA4:-NvaO69T1NfhxcHUSKA5:b-x6apjm) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/mod.rs#L106)):*
> ```XML
>             );
>             if peer.connection_id().is_none() {
>                 // in case we have a race condition where the connection is closed after we added to
> ```

TL;DR IMO you can erase this comment

I don't think that such a race condition can occur because we don't poll the behaviour and call `assign_peer_to_query` simultaneously, and from the NM's POV it doesn't matter if technically the connection closed happened before/during the `assign_peer_to_query`, the connection closed happened when it got the event about it
___
*[`crates/papyrus_network/src/peer_manager/mod.rs` line 109 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaP4EE-kLiiuiay-NI:-NvaP4EE-kLiiuiay-NJ:bdzrr14) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/mod.rs#L109)):*
> ```XML
>                 // the pending list the reciever will get an error and will need to ask for
>                 // re-assignment
>                 self.peer_pending_dial_with_event.insert(*peer_id, event);
> ```

Add a comment above that we're not connected to the peer and therefore we should dial to it
___
*[`crates/papyrus_network/src/peer_manager/mod.rs` line 109 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaPGTS9J54snxSPBdC:-NvaPGTS9J54snxSPBdD:bh4ww2k) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/mod.rs#L109)):*
> ```XML
>                 // the pending list the reciever will get an error and will need to ask for
>                 // re-assignment
>                 self.peer_pending_dial_with_event.insert(*peer_id, event);
> ```

What if there are multiple queries on this peer? Change the type of `peer_pending_dial_with_event` from `HashMap<K, V>` to `HashMap<K, Vec<V>>`
___
*[`crates/papyrus_network/src/peer_manager/peer.rs` line 26 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaR9Rg6WrlBGTaPKc5:-NvaR9Rg6WrlBGTaPKc6:bgy9crq) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/peer.rs#L26)):*
> ```XML
>     fn connection_id(&self) -> Option<ConnectionId>;
> 
>     fn set_connection_id(&mut self, connection_id: Option<ConnectionId>);
> ```

Consider `connection_id_mut` and returning a &mut to connection_id instead of receiving it
___
*[`crates/papyrus_network/src/peer_manager/test.rs` line 57 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaPqqOB62FoDonlD7q:-NvaPqqOB62FoDonlD7r:bd0b1z2) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/test.rs#L57)):*
> ```XML
>     // check assignment events
>     for event in peer_manager.pending_events {
>         if let Event::NotifyStreamedBytes(
> ```

Consider `let else { continue; };` in order to have 1 less indentation
___
*[`crates/papyrus_network/src/peer_manager/test.rs` line 344 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaRzMc7birjawCD7kV:-NvaRzMc7birjawCD7kW:blb0j0p) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/test.rs#L344)):*
> ```XML
> 
> #[tokio::test]
> async fn flow_test_assign_non_connected_peer() {
> ```

Add test for dial failure flow (or a TODO to add such test)
___
*[`crates/papyrus_network/src/streamed_bytes/behaviour.rs` line 341 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaSAh59Dq6muVvHV2t:-NvaSAh59Dq6muVvHV2u:b-fx90ft) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/streamed_bytes/behaviour.rs#L341)):*
> ```XML
>     // TODO: do something with the event. For now, we just ignore it.
>     fn on_other_behaviour_event(&mut self, event: mixed_behaviour::InternalEvent) {
>         if let mixed_behaviour::InternalEvent::NotifyStreamedBytes(internal_event) = event {
> ```

I think this is a worsening of the readability here but whatever you prefer

**Review by nagmo-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/papyrus/1908)* status: 1 of 5 files reviewed, 7 unresolved discussions (waiting on @ShahakShama)

___
*[`crates/papyrus_network/src/peer_manager/behaviour_impl.rs` line 18 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaNKXg6u4Fo0R5gZiC:-NvavH6m-exFovJtmn7E:b-896fix) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/behaviour_impl.rs#L18)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

This shouldn't be here. Here are events that reach the other behaviours or the NM
</blockquote></details>

Done.
___
*[`crates/papyrus_network/src/peer_manager/behaviour_impl.rs` line 96 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaM1dP5VqaHEYvAZ5W:-NvaY--W-cxeF4qakdGi:b-896fix) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/behaviour_impl.rs#L96)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

There might be multiple queries assigned to this peer. You need to re-assign all of them
</blockquote></details>

Done.
___
*[`crates/papyrus_network/src/peer_manager/behaviour_impl.rs` line 120 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaMz_8CYsTQ1SAyZze:-NvamcfN0eoz2VBjf6N3:b-896fix) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/behaviour_impl.rs#L120)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Check if the connection id of the peer is the closed connection
Also, add a TODO to support multiple connections for a single peer
</blockquote></details>

Done.
___
*[`crates/papyrus_network/src/peer_manager/mod.rs` line 106 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaO69T1NfhxcHUSKA4:-Nvap6qp7jJj-wtQd6uB:bae2aou) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/mod.rs#L106)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

TL;DR IMO you can erase this comment

I don't think that such a race condition can occur because we don't poll the behaviour and call `assign_peer_to_query` simultaneously, and from the NM's POV it doesn't matter if technically the connection closed happened before/during the `assign_peer_to_query`, the connection closed happened when it got the event about it
</blockquote></details>

I think it's an unlikely case to happen but if someone gets into it and tried to debug this comment would help so I think it makes sense to keep it
___
*[`crates/papyrus_network/src/peer_manager/mod.rs` line 109 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaP4EE-kLiiuiay-NI:-NvapQOI0Koo960VSV8J:b-48b0rs) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/mod.rs#L109)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Add a comment above that we're not connected to the peer and therefore we should dial to it
</blockquote></details>

it's obvious the line before says ` peer.connection_id().is_none()`
___
*[`crates/papyrus_network/src/peer_manager/mod.rs` line 109 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaPGTS9J54snxSPBdC:-NvaqOLQ6EAHLmClV_4k:b-896fix) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/mod.rs#L109)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

What if there are multiple queries on this peer? Change the type of `peer_pending_dial_with_event` from `HashMap<K, V>` to `HashMap<K, Vec<V>>`
</blockquote></details>

Done.
___
*[`crates/papyrus_network/src/peer_manager/test.rs` line 344 at r1](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvaRzMc7birjawCD7kV:-NvarzFz5Xocim0RkHaP:b-7i4imm) ([raw file](https://github.com/starkware-libs/papyrus/blob/d1a5e1da129cc4e35d4575109aff33d2cc59bb4a/crates/papyrus_network/src/peer_manager/test.rs#L344)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Add test for dial failure flow (or a TODO to add such test)
</blockquote></details>

there's a todo already

**Review by ShahakShama — APPROVED** [review]
:lgtm:

> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-:-Nvb_p1r3e-1wP9cnFSE:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @nagmo-starkware)

___
*[`crates/papyrus_network/src/peer_manager/behaviour_impl.rs` line 115 at r2](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvbXcnO78dPISQF3mnU:-NvbXcnO78dPISQF3mnV:bkezu8) ([raw file](https://github.com/starkware-libs/papyrus/blob/b81552b995f52ad0742746c7d4f3f0e15c1204ce/crates/papyrus_network/src/peer_manager/behaviour_impl.rs#L115)):*
> ```XML
>                         .get_mut(&peer_id)
>                         .expect(
>                             "in case we are waiting for a connection established event we assum \
> ```

Change "we assum" to "it's guaranteed that". If possible, add a short explanation why
___
*[`crates/papyrus_network/src/peer_manager/behaviour_impl.rs` line 132 at r2](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvbYBek9pV8aj-U9TXs:-NvbYBek9pV8aj-U9TXt:byaxrjb) ([raw file](https://github.com/starkware-libs/papyrus/blob/b81552b995f52ad0742746c7d4f3f0e15c1204ce/crates/papyrus_network/src/peer_manager/behaviour_impl.rs#L132)):*
> ```XML
>                         } else {
>                             error!(
>                                 "Connection closed event for a peer with a different connection \
> ```

Consider changing to "Peer has two connections open even though PeerManager ensures only 1 connection exists"
___
*[`crates/papyrus_network/src/peer_manager/mod.rs` line 112 at r2](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-Nvb_8y4Eg89ejrZYqCI:-Nvb_8y4Eg89ejrZYqCJ:bw5h61d) ([raw file](https://github.com/starkware-libs/papyrus/blob/b81552b995f52ad0742746c7d4f3f0e15c1204ce/crates/papyrus_network/src/peer_manager/mod.rs#L112)):*
> ```XML
>                 // the pending list, the reciever will get an error and will need to ask for
>                 // re-assignment
>                 if let Some(events) = self.peer_pending_dial_with_events.get_mut(peer_id) {
> ```

Consider using DefaultHashMap. see SB for example
probably not worth the short amount of time you have in papyrus

**Review by ShahakShama — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/papyrus/1908)* status: all files reviewed, 3 unresolved discussions (waiting on @nagmo-starkware)

___
*[`crates/papyrus_network/src/peer_manager/behaviour_impl.rs` line 132 at r2](https://reviewable.io/reviews/starkware-libs/papyrus/1908#-NvbYBek9pV8aj-U9TXs:-NvbajYT01GhAm-iwu9S:bqeuyby) ([raw file](https://github.com/starkware-libs/papyrus/blob/b81552b995f52ad0742746c7d4f3f0e15c1204ce/crates/papyrus_network/src/peer_manager/behaviour_impl.rs#L132)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Consider changing to "Peer has two connections open even though PeerManager ensures only 1 connection exists"
</blockquote></details>

Actually, now that I think of it. I think this will happen in the positive flow of the network because kademlia performs dials too. Consider changing the level of this message to debug
